crates/db_schema/src/impls/multi_community.rs

```diff
@@ -0,0 +1,127 @@
+use crate::{
+  newtypes::{CommunityId, MultiCommunityId, PersonId},
+  source::multi_community::{MultiCommunity, MultiCommunityInsertForm, MultiCommunityUpdateForm},
+  traits::Crud,
+  utils::{get_conn, DbConn, DbPool},
+};
+use diesel::{
+  dsl::{delete, insert_into},
+  update,
+  ExpressionMethods,
+  QueryDsl,
+};
+use diesel_async::RunQueryDsl;
+use lemmy_db_schema_file::schema::{multi_community, multi_community_entry};
+use lemmy_utils::error::LemmyResult;
+
+impl Crud for MultiCommunity {
+  type InsertForm = MultiCommunityInsertForm;
+  type UpdateForm = MultiCommunityUpdateForm;
+  type IdType = MultiCommunityId;
+
+  async fn create(pool: &mut DbPool<'_>, form: &MultiCommunityInsertForm) -> LemmyResult<Self> {
+    let conn = &mut get_conn(pool).await?;
+    Ok(
+      insert_into(multi_community::table)
+        .values(form)
+        .get_result(conn)
+        .await?,
+    )
+  }
+
+  async fn update(
+    pool: &mut DbPool<'_>,
+    id: MultiCommunityId,
+    form: &MultiCommunityUpdateForm,
+  ) -> LemmyResult<Self> {
+    let conn = &mut get_conn(pool).await?;
+    Ok(
+      update(multi_community::table.find(id))
+        .set(form)
+        .get_result(conn)
+        .await?,
+    )
+  }
+}
+
+impl MultiCommunity {
+  pub async fn upsert(conn: &mut DbConn<'_>, form: &MultiCommunityInsertForm) -> LemmyResult<Self> {
+    Ok(
+      insert_into(multi_community::table)
+        .values(form)
+        .on_conflict(multi_community::ap_id)
+        .do_update()
+        .set(form)
+        .get_result::<Self>(conn)
+        .await?,
+    )
+  }
+
+  pub async fn create_entry(
+    pool: &mut DbPool<'_>,
+    id: MultiCommunityId,
+    new_community: CommunityId,
+  ) -> LemmyResult<()> {
```

Comment:
This should probably be on the crud impl of the `MultiCommunityEntry` struct, and it should return `LemmyResult<Self>`